File: web/apps/app/lib/core/notes/note-export.ts
import type { Note } from '@oriveo/shared';
import { copyToClipboard } from '../../utils/clipboard';

function noteTitle(note: Note, untitled: string): string {
  return note.title.trim() || untitled;
}

function noteDate(note: Note): string {
  const date = new Date(note.createdAt);
  if (Number.isNaN(date.getTime())) return '';
  return date.toISOString().slice(0, 10);
}

function sourceLine(note: Note): string | null {
  if (note.captureKind === 'blank') return null;
  const parts = [note.sourceModelName, note.sourceProviderName, noteDate(note)]
    .filter((part): part is string => typeof part === 'string' && part.length > 0);
  if (parts.length === 0) return null;
  return `Source: ${parts.join(' - ')}`;
}

export function buildNoteMarkdown(note: Note, untitled: string): string {
  const sections: string[] = [`# ${noteTitle(note, untitled)}`];

  if (note.userNote?.trim()) {
    sections.push(note.userNote.trim().split('\n').map((line) => `> ${line}`).join('\n'));
  }

  if (note.tags.length > 0) {
    sections.push(`Tags: ${note.tags.join(', ')}`);
  }

  if (note.body.trim()) {
    sections.push(note.body.trim());
  }

  const source = sourceLine(note);
  const sourceParts = [source, note.sourcePrompt ? `Prompt: ${note.sourcePrompt}` : null]
    .filter((part): part is string => Boolean(part));
  if (sourceParts.length > 0) {
    sections.push(sourceParts.join('\n'));
  }

  return `${sections.join('\n\n')}\n`;
}

export function sanitizeNoteFilenamePart(value: string): string {
  const cleaned = value
    .replace(/[\/\\:*?"<>|]+/g, '-')
    .replace(/[\r\n]+/g, ' ')
    .replace(/\s+/g, ' ')
    .replace(/-+/g, '-')
    .trim()
    .replace(/^-+|-+$/g, '');
  return cleaned.slice(0, 80);
}

export function buildNoteMarkdownFilename(note: Note, untitled: string): string {
  const title = sanitizeNoteFilenamePart(noteTitle(note, untitled)) || untitled;
  const date = noteDate(note);
  return `${title}${date ? `-${date}` : ''}.md`;
}

export function downloadMarkdown(markdown: string, filename: string): void {
  const blob = new Blob([markdown], { type: 'text/markdown;charset=utf-8' });
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement('a');
  anchor.href = url;
  anchor.download = filename;
  document.body.appendChild(anchor);
  anchor.click();
  document.body.removeChild(anchor);
  URL.revokeObjectURL(url);
}

export function copyNoteMarkdown(note: Note, untitled: string): Promise<boolean> {
  return copyToClipboard(buildNoteMarkdown(note, untitled));
}

File: android/app/src/main/java/ai/oriveo/community/core/notes/NoteExport.kt
package ai.oriveo.community.core.notes

import ai.oriveo.community.core.model.Note
import ai.oriveo.community.core.model.NoteCaptureKind


object NoteExport {

    private val ILLEGAL_FILENAME = Regex("[/\\\\:*?\"<>|]+")
    private val NEWLINES = Regex("[\\r\\n]+")
    private val WHITESPACE = Regex("\\s+")
    private val DASHES = Regex("-+")
    private val TRIM_DASHES = Regex("^-+|-+$")

    private fun noteTitle(note: Note, untitled: String): String =
        note.title.trim().ifEmpty { untitled }

    private fun noteDate(note: Note): String = NoteTime.isoToDate(note.createdAt)

    private fun sourceLine(note: Note): String? {
        if (note.captureKind == NoteCaptureKind.Blank) return null
        val parts = listOfNotNull(
            note.sourceModelName?.takeIf { it.isNotEmpty() },
            note.sourceProviderName?.takeIf { it.isNotEmpty() },
            noteDate(note).takeIf { it.isNotEmpty() },
        )
        if (parts.isEmpty()) return null
        return "Source: ${parts.joinToString(" · ")}"
    }

    fun buildNoteMarkdown(note: Note, untitled: String): String {
        val sections = mutableListOf("# ${noteTitle(note, untitled)}")
        note.userNote?.trim()?.takeIf { it.isNotEmpty() }?.let { userNote ->
            sections.add(userNote.split("\n").joinToString("\n") { "> $it" })
        }
        if (note.tags.isNotEmpty()) sections.add("Tags: ${note.tags.joinToString(", ")}")
        note.body.trim().takeIf { it.isNotEmpty() }?.let { sections.add(it) }
        val sourceParts = listOfNotNull(
            sourceLine(note),
            note.sourcePrompt?.takeIf { it.isNotEmpty() }?.let { "Prompt: $it" },
        )
        if (sourceParts.isNotEmpty()) sections.add(sourceParts.joinToString("\n"))
        return sections.joinToString("\n\n") + "\n"
    }

    fun sanitizeNoteFilenamePart(value: String): String {
        val cleaned = value
            .replace(ILLEGAL_FILENAME, "-")
            .replace(NEWLINES, " ")
            .replace(WHITESPACE, " ")
            .replace(DASHES, "-")
            .trim()
            .replace(TRIM_DASHES, "")
        return cleaned.take(80)
    }

    fun buildNoteMarkdownFilename(note: Note, untitled: String): String {
        val title = sanitizeNoteFilenamePart(noteTitle(note, untitled)).ifEmpty { untitled }
        val date = noteDate(note)
        return if (date.isNotEmpty()) "$title-$date.md" else "$title.md"
    }
}
